### PR TITLE
ENH: Linalg: Python wrapper for ?geqrfp

### LINKED ISSUE #11169: Add wrapper for ?geqrfp
See [NumFOCUS Small Development Grant Proposal 2019 Round 3 proposal](https://www.dropbox.com/s/i24h3j2osol56uf/Small%20Development%20Grant%20-%20LAPACK%20%20%28Condensed%29.pdf?dl=0) for background.

Add wrapper for [?geqrfp](http://www.netlib.org/lapack/explore-html/dd/d9a/group__double_g_ecomputational_ga5dc72a0f55ff4c08f1a3596f56e04aaa.html).

Existing [wrapper](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lapack.dgeqrf.html) for [?geqrf](http://www.netlib.org/lapack/explore-html/df/dc5/group__variants_g_ecomputational_ga3766ea903391b5cf9008132f7440ec7b.html) provides some guidance.

Suggested signature:
```
! rq_a,tau,work,info = geqrfp(a,lwork=3*n,overwrite_a=0)
```
We will also want to create a routine to compute the optimal `lwork` value:
```
! work,info = geqrfp_lwork(n)
```

I'm a bit confused about how `overwrite_a` works - is that optional argument added automatically by F2PY? It is an optional argument for `?geqrf` but there's no "code" for it in the wrapper.
_Update: See @ilayn's response below._

If the user provides `lwork=-1`, it seems that `?geqrfp` is supposed to determine the optimal `lwork` rather than actually solving the problem. I suppose the idea is that if you don't know what to choose for `lwork`, you call  `?geqrfp` once with `lwork=-1` just to get the optimal value of `lwork`, then call it again with that value of `lwork` to actually solve the problem. Is that right? Should the wrapper go through that process if no `lwork` is provided by the user? The wrapper for `?geqrf` doesn't use this approach, it seems; it just defaults to 3*n. Where does 3*n come from? In the [NAG manual](https://www.nag.com/numeric/fl/nagdoc_latest/html/f08/f08aef.html), it looks like 64*n is used.
_Update: See @ilayn's response below._

As for tests, the [NAG manual](https://www.nag.com/numeric/fl/nagdoc_latest/html/f08/f08aef.html) provides an example for `?geqrf`, so these could be adapted for `?geqrfp`. Alternatively, we could follow the approach used by [`scipy.linalg.qr`](https://github.com/scipy/scipy/blob/master/scipy/linalg/decomp_qr.py) to convert the output into matrices `Q` and `R` and write tests based on those, e.g.

- Create random matrix `A`
- Decompose using wrapper
- Convert to `Q` and `R` using approach in `scipy.linalg.qr`
- Verify that `R` is upper triangular and has all non-negative diagonal 
- Verify that `Q` is orthogonal
- Verify that `A` = `Q @ R`

etc.
Probably should do this for an example where regular QR gives negative entries on diagonal of `R`. We should also check for empty matrix and both tall and wide matrices as input.

### PR DESCRIPTION
**Pull request by Kai-Striega**
This PR adds a python wrapper for the set of LAPACK functions [`?geqrfp`](http://www.netlib.org/lapack/explore-html/dd/d9a/group__double_g_ecomputational_ga5dc72a0f55ff4c08f1a3596f56e04aaa.html) as part of the NumFoucs small development grant. ?geqrfp computes the QR decomposition of a real general matrix with non-negative elements on the diagonal of R.  This will close #11169.  

Again special thanks to @swallan and @mdhaber for their contributions!

Note: this PR closely duplicates a helper function ``generate_random_dtype_array`` function used in #11376. To avoid duplicate functions I intend to fix it as part of the merge commit once that PR has been merged.

### CODE REVIEW

File: scipy/linalg/flapack_gen.pyf.src [hunk 1]
```diff
@@ -990,12 +990,58 @@ subroutine <prefix>geqrf_lwork(m,n,a,tau,work,lwork,info)
...
+    !    entries;
+    !    0 is a (M-N)-by-N zero matrix, if M > N.
+
+    callstatement (*f2py_func)(&m,&n,a,&lda,tau,work,&lwork,&info)
+    callprotoargument int*,int*,<ctype>*,int*,<ctype>*,<ctype>*,int*,int*
+
+    integer intent(hide), check(m > 0), depend(a) :: m = shape(a, 0)
+    integer intent(hide), check(n > 0), depend(a) :: n = shape(a, 1)
+    <ftype> intent(in,out,copy,out=qr), dimension(m,n) :: a
+    integer intent(hide) :: lda = max(1, m)
```
Comment (ilayn) [inline scipy/linalg/flapack_gen.pyf.src]: This needs a depend clause but better keep it explicit with `shape(a, 0)` such that f2py doesn't need to resolve any dependency relation.

File: scipy/linalg/flapack_gen.pyf.src [hunk 2]
```diff
@@ -990,12 +990,58 @@ subroutine <prefix>geqrf_lwork(m,n,a,tau,work,lwork,info)
...
+    <ftype> intent(out), depend(m,n), dimension(MIN(m,n)) :: tau
+    <ftype> intent(out), depend(lwork), dimension(lwork) :: work
```
Comment (ilayn) [inline scipy/linalg/flapack_gen.pyf.src]: We don't need `work` to be exposed, it is needed for the lwork computation

File: scipy/linalg/flapack_gen.pyf.src [hunk 3]
```diff
@@ -990,12 +990,58 @@ subroutine <prefix>geqrf_lwork(m,n,a,tau,work,lwork,info)
...
+    ! work, info = geqrfp_lwork(m, n)
+
+    fortranname <prefix>geqrfp
+    callstatement (*f2py_func)(&m,&n,&a,&lda,&tau,&work,&lwork,&info)
+    callprotoargument int*,int*,<ctype>*,int*,<ctype>*,<ctype>*,int*,int*
+
+    integer intent(in), check(m > 0) :: m
+    integer intent(in), check(n > 0) :: n
+    <ftype> intent(hide) :: a
+    integer intent(hide) :: lda = m
```
Comment (ilayn) [inline scipy/linalg/flapack_gen.pyf.src]: depends on `m`

File: scipy/linalg/tests/test_lapack.py [hunk 4]
```diff
@@ -1654,3 +1655,95 @@ def test_getc2_gesc2():
...
+    if dtype in COMPLEX_DTYPES:
+        return (np.random.rand(*shape)
+                + np.random.rand(*shape)*1.0j).astype(dtype)
+    return np.random.rand(*shape).astype(dtype)
+
+
+@pytest.mark.parametrize('dtype', DTYPES)
+@pytest.mark.parametrize('matrix_size', [(3, 4), (7, 6), (6, 6)])
+def test_geqrfp(dtype, matrix_size):
+    # Tests for all dytpes, tall, wide, and square matricies.
```
Comment (ilayn) [inline scipy/linalg/tests/test_lapack.py]: nitpick: matricies -> matrices

File: scipy/linalg/tests/test_lapack.py [hunk 5]
```diff
@@ -1654,3 +1655,95 @@ def test_getc2_gesc2():
...
+    # Using the routine with random matrix A, Q and R are obtained and then
+    # tested such that R is upper triangular and non-negative on the diagonal,
+    # and Q is an orthagonal matrix. Verifies that A=Q@R. It also
+    # tests against a matrix that for which the  linalg.qr method returns
+    # negative diagonals, and for error messaging.
+
+    # Additional note: this test stands alone without a NAG manual additional
```
Comment (ilayn) [inline scipy/linalg/tests/test_lapack.py]: I think you can skip this note

File: scipy/linalg/tests/test_lapack.py [hunk 6]
```diff
@@ -1654,3 +1655,95 @@ def test_getc2_gesc2():
...
+    assert_allclose(np.eye(q.shape[0]), q@(q.conj().T), rtol=rtol,
+                    atol=atol)
+    # ensure r is upper tri by comparing original r to r as upper triangular
+    assert_allclose(r, np.triu(r), rtol=rtol)
+    # make sure diagonals of r are positive for this random solution
+    assert_(np.all(np.diag(r) > np.zeros(len(np.diag(r)))))
+    # ensure that info is zero for this success
+    assert_(info == 0)
+
+    # test that this routine gives r diagonals that are positive for a
```
Comment (ilayn) [inline scipy/linalg/tests/test_lapack.py]: Maybe split this and the empty array to a separate test for better error handling?

**Review by ilayn — CHANGES_REQUESTED** [review]

File: scipy/linalg/flapack_gen.pyf.src (hunk 1 above)
Comment (Kai-Striega) [reply scipy/linalg/flapack_gen.pyf.src]: So that should be: ``integer intent(hide), depend(a)  :: lda = max(1,shape(1, 0))``?

Comment (ilayn) [reply scipy/linalg/flapack_gen.pyf.src]: Yes indeed with `shape(a, 0)`

**Review by ilayn — APPROVED** [review]

**Comment by Kai-Striega** [comment]
@ilayn Thanks for your patience! If there are no other changes and the
tests pass, would you be willing to merge it?

**Comment by ilayn** [comment]
Oh I didn't want to step on you folks' toes and tried to silently get out of the way but sure, clicking now.

**Comment by Kai-Striega** [comment]
Thanks! It feels good to have the first one done.

**Comment by Kai-Striega** [comment]
And also congratulations to @swallan for getting his first set of commits merged into SciPy!

**Comment by ilayn** [comment]
Keep 'em coming @swallan !

**Comment by mdhaber** [comment]
Thanks @swallan! If you'd like your name added to [THANKS.txt](https://github.com/scipy/scipy/blob/master/THANKS.txt), please make an entry [here](https://github.com/scipy/scipy/wiki/*THANKS.txt-additions-modifications-for-1.5.0*).

**Pull request merged**